### ISSUE TITLE
Is it possible to implement ipython magics like in `line_profile`?

### ISSUE THREAD

**Issue by magonser** [vpelletier/pprofile#21]
Hello,

thanks for pprofile! I don't know much about internals of iPython or pprofile, but I've noticed, that line_profiler can be loaded within a iPython notebook via `%load_ext line_profiler` and subsequently being used in iPython. My question: is this possible for pprofile too and/or it's implementation planned?

Details on how line_profile does it, can be seen here:
https://github.com/rkern/line_profiler/blob/master/line_profiler.py

Regards, Markus

**Comment by vpelletier** [comment]
Hello, thanks for your interest !

My problem with implementing this is that I do not use ipython (notebook or otherwise) enough to understand exactly how it would be used, like what deviates from normal module usage.

If you have ideas about how it would be best integrated, usage examples (mock-up or more precise), it could help me getting into it.

**Comment by flothesof** [comment]
Hi @vpelletier,

I would also love to see a line_magic or a cell magic appear. 
I can give you a use case since I just wrote a blog post about profiling and optimizing: [link](http://flothesof.github.io/optimizing-python-code-numpy-cython-pythran-numba.html).

My use case is: I have a function that I am trying to speed up. To do that, I need to know what takes time. So  I run controlled tests on a benchmark. Say:
```python
import numpy as np
image = np.random.rand(200, 300)
def laplace_numpy(image):
    """Laplace operator in NumPy for 2D images."""
    out = image[:-2, 1:-1] + image[2:, 1:-1] + image[1:-1, :-2] + image[1:-1, 2:] - 4*image[1:-1, 1:-1]
    return np.abs(out) > 0.05
benchmark = lambda : laplace_numpy(image)
```
Now, I would like to know where my function spends time. So I would run something like this:
`%pprofile benchmark()`
And I would get the output in my jupyter notebook cell (see the profiling and timing section in the blog article).
This would be a terrific addition to pprofile. 

Best regards,
Florian

**Comment by vpelletier** [comment]
@flothesof thanks for bumping this topic. Actually, there was already a shorter way to do this - although still not a line/cell magic:

```python
pprofile.run('foo()')
```

Nevertheless, I just pushed a few commits which declares `%pprofile` and `%%pprofile`, and which I tested under jupyter notebook 4.2.3 with python3.

For now, `%profile` is rather dumb: merely syntactic sugar for `pprofile.run`. I wonder if it would make sense to start parsing arguments - and which ones. Annotation does not seem to work very well (blank lines, several cells getting overlapped so final result of non-trivial profiling is unreadable).

`%%pprofile` OTOH behaves like pprofile command line, accepting all usual parameters, with the python executable being replaced by cell content. Annotation is working a bit better: called cells are separated in annotated output, but `%%pprofile` cell is not.

**Comment by vpelletier** [comment]
Released as 1.11.0 . I kept the magics undocumented as I feel I am not making good enough use of jupyter capabilities, so their invocation is likely to evolve.

For example, I think profiling output could become some interactive view of the annotated code, kind of a simpler kcachegrind in a jupyter cell.
